### ISSUE TITLE
DataGrid - integrated filter builder

### ISSUE THREAD

**Issue by dxbykov** [DevExpress/DevExtreme#3565]
# The Problem

* It is impossible to build complex filtering criteria using the DataGrid filter row or column filter. The FilterBuilder component allows this but requires extra coding to bind it to a DataGrid. 
* Sometimes it's convenient to have a 'long-term' filter applied and then perform a 'short-term' extra filtering on the filtered rows. It’s hard to do with current API.

# The Solution

We integrated the DataGrid and FilterBuilder components in v18.1. The DataGrid renders the current filter criteria in the new filter panel. You can use the following configuration to enable and control this feature:

```js
  filterPanel: {
    visible: true,  // show/hide filter panel
    filterEnabled: true // enable/disable filtering
  },
  filterValue: [], // set/get current filter criteria
  filterSyncEnabled: 'auto' // enable/disable filtering UI elements synchronization
```
Refer to the [filtering](https://js.devexpress.com/Documentation/18_1/Guide/Data_Layer/Data_Layer/#Reading_Data/Filtering) guide for more information about the `filterValue` option format.

## Filters Synchronization

The DataGrid allows a user to filter data using a filter row, column filter, search panel and filter panel. It’s possible to merge the criteria from all DataGrid filtering controls and show the combined filter expression in the filter panel. The `filterSyncEnabled` option mentioned above controls this behavior - its default value is `'auto'`. In this mode, synchronization is enabled when the filter panel is visible.

# Try It And Give Us Feedback 

## Live Sandbox
[jQuery](https://codepen.io/dxbykov/pen/zWJMjG) 
[Angular](https://embed.plnkr.co/BHxpFg7gDEBfE7Vg2Uw6/)

## Installation

Follow the installation guide from the [release page](https://github.com/DevExpress/DevExtreme/releases) and leave us feedback on this topic.

**Comment by genachka** [comment]
While I do like very much that you added the ability to have additional filtering, as your current column filtering leaves A LOT to be desired, I'm not a fan of having a dedicated row on the bottom and having to have two experiences for filtering. Really wish the column filtering was more extensive.

Having used JQWidgets jqxGrid in the past, really miss the default ability they have in their column filtering. They show an icon on column filter when filter is enabled and it's quick and easy to add / change filtering:
![jqxgrid column filtering](https://user-images.githubusercontent.com/7947294/38457979-fad0175c-3a65-11e8-8766-1b048e7cdd1d.png)

**Comment by dxbykov** [comment]
@genachka Thanks for sharing your thoughts. We'll take them into consideration.

**Comment by FabianoCastro** [comment]
Hi, I had tried it, but the PanelFilter does not update after restore the state of the DataGrid.

Thanks
![2018-04-26 23_55_12-crm_retornos listagem - is gestao](https://user-images.githubusercontent.com/34251113/39342232-663d69bc-49ad-11e8-8fdb-2a343a94d6c9.png)

**Comment by komarovalexander** [comment]
@FabianoCastro thank you for your feedback.
I've reproduced this behavior and created a corresponding [ticket](https://www.devexpress.com/Support/Center/Question/Details/T630533/datagrid-shows-old-filter-values-but-does-not-apply-them-when-the-filtersyncenabled) in our Support Center. You can subscribe to it to receive a notification of our progress.

**Comment by dxbykov** [comment]
Thanks to everybody who participated in this discussion or gave us early feedback by other means. The feature is now available in the [v18.1](https://github.com/DevExpress/DevExtreme/releases/tag/18.1.3) release. I'm closing this thread. In the case of bugs or questions, feel free to create new GitHub [issues](https://github.com/DevExpress/DevExtreme/issues) or tickets in our [Support Center](https://www.devexpress.com/Support/Center/).
